### PR TITLE
tetragon: Add bench application

### PR DESCRIPTION
**Pull request by olsajiri**
Adding benchmark application to measure tracing workloads. 

The tetragon-bench application starts tetragon (or not for -baseline option)
for given CRD file and runs workload.

At the end it displays some stats, like:

```
Benchmark summary
-----------------
Test started:      2022-09-08 15:04:28.04513503 +0200 CEST m=+0.016539427
Test ended:        2022-09-08 15:04:38.268072909 +0200 CEST m=+10.239477312
Workload start:    2022-09-08 15:04:38.268072909 +0200 CEST m=+10.239477312
Arguments:         
Total duration:    10.222937885s
Setup duration:    1.743823005s
Workload duration: 8.479094472s
Test duration:     10.222937885s
Export duration:   0s
Export stats:      bytes=0, writes=0
FGS cpu usage:     system=15.170223s, user=17.522276s, rss=230504, ctxsw=1231909
Ring buffer:       received=3958406, lost=41842, errors=0
BPF statistics:
  event_execve                   [tracepoint/212]   :   8.89µs (2)
  event_wake_up_new_task         [kprobe/211]       :   1.41µs (6)
  generic_kprobe_event           [kprobe/205]       :   1.54µs (4001580)
  restrict_filesystems           [lsm/58]           :   0.13µs (1569)
  event_exit                     [tracepoint/210]   :   0.49µs (32)
```

It's possible to run predefined workloads with builtin CRD files,
or run specific custom benchmark.

The predefined benchmarks are rw and open and can be run with:

```
# ./tetragon-bench -trace rw
# ./tetragon-bench -trace open
```
They do and monitor some simple read/write open operations and they are there
mainly to show the possibility of builtin benchmark if we think it's good idea ;-)

The custom benchmark is what I'm using most at the moment and
it allows to pass CRD file to start tetragon with and specify workload, like:

```
# ./tetragon-bench -crd crds/examples/names.yaml  -- perf bench sched pipe
```

I'm doing perf profiles to check on CRD stuff with:

```
# perf record -g -e cycles:k ./tetragon-bench -crd ./read_write.yaml -json-encode -- perf bench sched pipe
```

Signed-off-by: Jiri Olsa <jolsa@kernel.org>

### CODE REVIEW

File: pkg/bench/bench.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+	} else {
+		exePath, err := os.Executable()
+		if err != nil {
+			log.Fatal(err)
+		}
+		option.Config.HubbleLib = path.Join(path.Dir(exePath), "bpf/objs")
+
+		if _, err := os.Stat(option.Config.HubbleLib); err != nil {
+			// Running outside the source tree, fall back to default location.
+			option.Config.HubbleLib = "/var/lib/hubble-fgs"
```
Comment (tixxdz) [inline pkg/bench/bench.go:93]: Maybe this one: https://github.com/cilium/tetragon/blob/main/pkg/defaults/defaults.go#L44 ?

File: pkg/bench/bpfstats.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"time"
+)
+
+func EnableBpfStats() {
+	err := os.WriteFile("/proc/sys/kernel/bpf_stats_enabled", []byte("1"), 0666)
```
Comment (tixxdz) [inline pkg/bench/bpfstats.go:16]: Is the perm mode ok? maybe restrict it a bit

File: pkg/bench/cpuusage.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	cpuUsage.SystemTime = timevalToDuration(rusage.Stime)
+	cpuUsage.MaxRss = rusage.Maxrss
+	cpuUsage.ContextSwitches = rusage.Nivcsw + rusage.Nvcsw
+	return
+}
+
+func CPUUsageFromCPUAcct(containerID string) CPUUsage {
+
+	rss := int64(0)
+	memStatFilename := fmt.Sprintf("/sys/fs/cgroup/memory/docker/%s/memory.stat", containerID)
```
[the remark this answers is not in the document]
Comment (tixxdz) [reply pkg/bench/cpuusage.go:41]: btw where this is called?

[the remark this answers is not in the document]
Comment (tixxdz) [reply pkg/bench/cpuusage.go:41]: The full cgroup path may not container docker directory at all or another container runtime, I'm missing context, but if you could read /proc/self/cgroups and adapt according to cgroup version and the controllers you are interested in to get right working paths without hardcoding it here. There is some logic in that cgroups.go package

File: pkg/bench/bench.go (hunk 1 above)
Comment (olsajiri) [reply pkg/bench/bench.go:93]: ok, will use that one, thanks

File: pkg/bench/bpfstats.go (hunk 2 above)
Comment (olsajiri) [reply pkg/bench/bpfstats.go:16]: that permission will be used only if the file does not exist and needs to be created, so we are fine here

File: pkg/bench/cpuusage.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (olsajiri) [reply pkg/bench/cpuusage.go:41]: right, I'll remove it for now, because it's not used and check on that more, thanks

File: pkg/bench/summary.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+	"github.com/cilium/tetragon/pkg/metrics/ringbufmetrics"
+	"github.com/fatih/color"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+// Summary gathers benchmark results. Serializes to JSON.
+// This is updated from multiple places concurrently, but currently
+// there is no overlap on writes, so this isn't yet protected by a mutex.
```
Comment (kkourt) [inline pkg/bench/summary.go:21]: Are the concurrent writes all happen in BpfStats?

File: pkg/bench/summary.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	RunTime            time.Time
+	SetupDurationNanos time.Duration
+	TestDurationNanos  time.Duration
+
+	JSONEncodingDurationNanos time.Duration
+	ExportStats               CountingDiscardWriter
+
+	TetragonCPUUsage CPUUsage
+
+	BpfStats map[int64]*BpfProgStats
```
Comment (kkourt) [inline pkg/bench/summary.go:38]: Is CPU id the key here?

File: pkg/bench/bench.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	cnf, err := config.ReadConfigYaml(string(yamlData))
+	if err != nil {
+		return nil, err
+	}
+
+	return cnf, nil
+}
+
+func (args *Arguments) String() string {
+	return fmt.Sprintf("")
```
Comment (kkourt) [inline pkg/bench/bench.go:70]: Is this intentional?

Maybe something like `fmt.Sprintf("%+v", args)` would be better?

File: pkg/bench/cpuusage.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+func (cu CPUUsage) Sub(cu2 CPUUsage) CPUUsage {
+	cu.UserTime -= cu2.UserTime
+	cu.SystemTime -= cu2.SystemTime
+	cu.ContextSwitches -= cu2.ContextSwitches
+	return cu
+}
+
+func (cu CPUUsage) Add(cu2 CPUUsage) CPUUsage {
+	cu.UserTime += cu2.UserTime
+	cu.SystemTime += cu2.SystemTime
```
Comment (kkourt) [inline pkg/bench/cpuusage.go:58]: `Sub` has a `ContextSwitches` as well? maybe it's needed here as well?

File: pkg/bench/trace.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import (
+	"context"
+	"log"
+	"os"
+	"strings"
+	"time"
+)
+
+type TraceBench interface {
+	Crd(ctx context.Context, args *Arguments) string
```
Comment (kkourt) [inline pkg/bench/trace.go:15]: What is this method supposed to do? Maybe we can come up with a more informative name?

Based on the caller, it returns the configuration filename? Maybe `ConfigFilename()`?

File: Makefile [hunk 9]
```diff
@@ -75,13 +75,16 @@ tetragon-bpf-container:
 verify: tetragon-bpf
 	sudo contrib/verify/verify.sh bpf/objs
 
-.PHONY: tetragon tetra tetragon-operator tetragon-alignchecker
+.PHONY: tetragon tetra tetragon-operator tetragon-alignchecker bench
```
Comment (kkourt) [inline Makefile:78]: How about we name the binary `tetragon-bench`?

**Review by kkourt — CHANGES_REQUESTED** [review]
Thanks! Some minor requests / suggestions.

File: Makefile (hunk 9 above)
Comment (kkourt) [reply Makefile:78]: Could we also add the binary to `.gitignore`?

Comment (olsajiri) [reply Makefile:78]: yes, good idea, thanks

File: pkg/bench/trace.go (hunk 8 above)
Comment (olsajiri) [reply pkg/bench/trace.go:15]: right, it gives back the CRD file path. will rename that

File: pkg/bench/cpuusage.go (hunk 7 above)
Comment (olsajiri) [reply pkg/bench/cpuusage.go:58]: good catch, it's a bug, will fix

File: pkg/bench/bench.go (hunk 6 above)
Comment (olsajiri) [reply pkg/bench/bench.go:70]: leftover, will add, thanks

File: pkg/bench/summary.go (hunk 4 above)
Comment (olsajiri) [reply pkg/bench/summary.go:21]: it's the ExportStats that are updated from tetragon process,
rest should be done just in bench process, I'll update the comment

File: pkg/bench/summary.go (hunk 5 above)
Comment (kkourt) [reply pkg/bench/summary.go:38]: Is see. Could you please add a comment about what the key is?

Comment (olsajiri) [reply pkg/bench/summary.go:38]: yep, will do

**Review by kkourt — APPROVED** [review]
Thanks!

**Pull request merged**